### PR TITLE
[617E] Limit the size of configuration files in UI editor

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftersoftware/craftercms/issues/617

### CODE REVIEW

File: ui/app/src/utils/content.ts [hunk 1]
```diff
@@ -77,6 +77,8 @@ import { showCodeEditorDialog, showEditDialog } from '../state/actions/dialogs';
 import { Dispatch } from 'react';
 import { AnyAction } from 'redux';
 
+export const MAX_CONFIG_SIZE = 512000; // Max configuration files allowed (in bytes)
```
Comment (rart) [inline ui/app/src/utils/content.ts]: Move to constants.

**Review by rart — CHANGES_REQUESTED** [review]
What was the delta of the calculation in UI vs the OS?

**Comment by rart** [comment]
What's the answer to the question above?

**Comment by jvega190** [comment]
@rart exact max size in bytes for configuration files used in backend is 512*1024, differing from the OS calculation (524288 studio vs 512000 OS). Tested with the exact max size in bytes set by backend and when using an extra char/byte the backend request fails.

File: ui/app/src/utils/constants.ts [hunk 2]
```diff
@@ -99,3 +99,5 @@ export const XSRF_TOKEN_COOKIE_NAME = 'XSRF-TOKEN';
 export const plugins = new Map<string, PluginDescriptorWithSource>();
 
 export const components = new Map<string, WidgetRecord>();
+
+export const MAX_CONFIG_SIZE = 512 * 1024; // Max configuration files allowed (in bytes)
```
Comment (rart) [inline ui/app/src/utils/constants.ts]:
```suggestion
export const MAX_CONFIG_FILE_SIZE = 524288; // Max configuration file size allowed (in bytes) (512 * 1024 = 524288)
```

File: ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx [hunk 3]
```diff
@@ -122,8 +126,18 @@ export function GlobalConfigManagement() {
...
+      dispatch(
+        showSystemNotification({
+          message: formatMessage({
+            defaultMessage: 'Maximum configuration size exceeded, please reduce it in order to properly save.'
+          }),
+          options: {
+            variant: 'error'
+          }
+        })
+      );
```
Comment (rart) [inline ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx]: With the [new approach](https://github.com/craftersoftware/craftercms/issues/617#issuecomment-1813948305) in mind, perhaps we simply disable the Save button and we don't need this.

File: ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx [hunk 4]
```diff
@@ -106,6 +108,8 @@ export function GlobalConfigManagement() {
   };
 
   const onSaveClick = () => {
+    const value = aceEditorRef.current.getValue();
+    const contentSize = getContentSizeInBytes(value);
```
Comment (rart) [inline ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx]: Let's just match the backend and check for string length.

File: ui/app/src/utils/content.ts [hunk 5]
```diff
@@ -1067,3 +1067,7 @@ export function generateComponentBasePath(contentType: string) {
 export function generateComponentPath(modelId: string, contentType: string) {
   return `${generateComponentBasePath(contentType)}/${modelId}.xml`;
 }
+
+export function getContentSizeInBytes(content: string) {
+  return new TextEncoder().encode(content).length;
+}
```
Comment (rart) [inline ui/app/src/utils/content.ts]: Guess we don't need this anymore.

**Review by rart — CHANGES_REQUESTED** [review]
See [comment on the ticket](https://github.com/craftersoftware/craftercms/issues/617#issuecomment-1813948305).

File: ui/app/src/utils/content.ts (hunk 5 above)
Comment (rart) [reply ui/app/src/utils/content.ts]: Not resolved.

File: ui/app/src/utils/constants.ts (hunk 2 above)
Comment (rart) [reply ui/app/src/utils/constants.ts]: Not resolved.

File: ui/app/src/utils/content.ts (hunk 5 above)
Comment (jvega190) [reply ui/app/src/utils/content.ts]: Committed

File: ui/app/src/utils/constants.ts (hunk 2 above)
Comment (jvega190) [reply ui/app/src/utils/constants.ts]: Just committed

**Comment by jvega190** [comment]
@rart tested with the max config size and there is no noticeable performance downgrade

**Review by rart — APPROVED** [review]

**Pull request merged**